### PR TITLE
Reimplement 2d grid reorientation

### PR DESCRIPTION
**Pull request by bangerth**
This patch has been 13 years coming, and most of it was implemented over a year ago.

The patch re-implements the algorithm we use to orient edges in 2d in such a way that the coordinate system on each edge matches the restriction of the coordinate system on the cell to that edge. The old implementation, written in 2003 by Mike Anderson, is quite concise and of high quality. But it is not extensively documented, and is unnecessarily hard to read because it uses different conventions for names etc than we use in the rest of the library.

This reimplementation uses standard conventions, and it uses the language and symbols of a paper currently in review and for which a preprint is available at
  http://arxiv.org/abs/1512.02137
This paper concisely describes the algorithms and data structures used in this algorithm. As an additional benefit, it uses the standard vertex ordering of cells rather than the old-style ordering used in the existing algorithm.

Following the algorithm without understanding the graph theoretical context of the problem may be difficult, but because it uses the same symbols as in the paper, should be easy enough if you have the paper.

There is really not very much to the algorithm once you understand how it is written in the paper. The only tricky part is that I wanted to make it compatible to the existing implementation. After a good deal of debugging, I came to realize that the old implementation had a bug in it that accidentally flipped every edge. This is easy to accommodate, so I incorporated the relevant change into this patch in the form of the second commit, leaving the first of the two commits to just cover the basic implementation of the algorithm as described in the paper.

I've run the patch through the testsuite and found that it does not cause any regressions. The first commit alone, i.e., without the backward compatibility fix discussed in the previous paragraph, did cause a dozen or two tests to fail. I conclude that the code's effects are sufficiently well covered in the testsuite to not need more testcases.

### CODE REVIEW

**Comment by bangerth** [comment]
@mikeando: FYI

**Comment by bangerth** [comment]
I should also say that I hope to eventually run the 3d case through the same code paths, with minor modifications to the data structures. This would save several 100 lines of code. But that's for another day.

**Comment by bangerth** [comment]
Also relevant: this is a complete re-implementation of basic functionality, but the github diff view still tries to show the result as a diff, which makes it really confusing to read the patch because there is no correspondence between the plusses and minusses github shows. You're likely better off looking at the split diff view:
https://github.com/dealii/dealii/pull/3166/files?diff=split

Alternatively, just look at the state of the code introduced by this patch alone, without reference to the old code. The new code is the complete contents of `namespace internal::GridReordering2d`.

**Comment by mikeando** [comment]
After a quick reread of the old code, and some dredging of my memory banks, I think some of the way the original code was written was to ensure we never needed to make a copy of the mesh, only modify the existing one in memory. 

A quick glance over the new version suggests that it recreates the entire mesh using the new data structures and orients that. 

That change is probably OK, but could be (a) slower, (b) more memory hungry.

But as you say that code is 13 years old, so its likely my memory of it is fuzzy at best.

**Comment by bangerth** [comment]
I think the opposite is actually true :-) My `Edge` structure corresponds almost 1:1 to the `MEdge` class, but my `Cell` structure only stores the indices of vertices and edges whereas you built the `MQuad` structure that not only has vertices and edges, but in fact the whole `CellData` object as a member variable. I think in the end, we may even save a bit of memory.

I am ultimately not concerned about memory because at the time when we run this algorithm, there are no `Triangulation`, `DoFHandler`, or `SparseMatrix` objects around. These are so vastly more memory consuming that I'm really not worried about memory here :-)

**Comment by bangerth** [comment]
For the record, creating the separate edge and cell data structures happened in lines 349 and 374 of the old code, in the `build_graph()` function.

**Comment by mikeando** [comment]
Yep. I was misinterpreting an out reference on one of my old functions. What an abomination. Thank goodness C++ doesn't need that kind of cruft anymore - I think your version is much cleaner.

File: source/grid/grid_reordering.cc [hunk 1]
```diff
@@ -29,585 +29,710 @@
...
-     * Simple data structure denoting
-     * an edge, i.e. the ordered pair
-     * of its vertices. This is only
-     * used in the is_consistent
+     * A simple data structure denoting an edge, i.e., the ordered pair
+     * of its vertex indices. This is only used in the is_consistent()
      * function.
      */
-    struct Edge
+    struct CheapEdge
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: In my opinion a struct without public data members should be a class.

File: source/grid/grid_reordering.cc [hunk 2]
```diff
@@ -29,585 +29,710 @@
...
-    }
+      OrientationStatus orientation_status;
 
+      /**
+       * A structure that stores how this edge relates to the at most
+       * two cells that are next to it.
+       */
+      struct AdjacentCell
+      {
+        unsigned int cell_index;
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: You could add a constructor that makes this struct invalid and rip out the initialization you do above.

File: source/grid/grid_reordering.cc [hunk 3]
```diff
@@ -29,585 +29,710 @@
...
+        for (unsigned int i=0; i<GeometryInfo<dim>::vertices_per_cell; ++i)
+          vertex_indices[i] = numbers::invalid_unsigned_int;
+        for (unsigned int i=0; i<GeometryInfo<dim>::lines_per_cell; ++i)
+          edge_indices[i] = numbers::invalid_unsigned_int;
       }
 
-      bool side_sort_less(const MSide &s1, const MSide &s2)
+      /**
+       * Construct a Cell object from a CellData object. Also take a (sorted)
+       * list of edges and to point into from the current object.
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: remove "and"? Or I don't understand what you are trying to say.

File: source/grid/grid_reordering.cc [hunk 4]
```diff
@@ -29,585 +29,710 @@
...
+       * @param c
+       * @param edge_list
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: remove if you are not going to explain them.

File: source/grid/grid_reordering.cc [hunk 5]
```diff
@@ -29,585 +29,710 @@
...
+          if (edges[cells[c].edge_indices[l]].orientation_status == Edge<dim>::not_oriented)
+            return c;
+
+      return numbers::invalid_unsigned_int;
     }
 
 
+    /**
+     * Given a set of cells and edges, orient all edges that are
+     * (globall) parallel to the one identified by the @p cell and
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: "global"?

File: source/grid/grid_reordering.cc [hunk 6]
```diff
@@ -29,585 +29,710 @@
...
+
+      // choose the direction of the first edge. we have free choice
+      // here and could simply choose "forward" if that's what pleases
+      // us. however, for backward compatibility with the previous
+      // implementation used till 2016, let us just choose the
+      // direction so that it matches what we have in the given cell.
+      //
+      // in fact, in what can only be assumed to be a bug in the
+      // original implementation, after orienting all edges, the code
+      // that rotates the cells so that they match edge orientations
```
Comment (tjhei) [inline source/grid/grid_reordering.cc]: I don't understand the point. If it is a bug, why do you still keep this behavior? Or if you fixed the bug, there is no reason to talk about it here anymore, right?

**Comment by bangerth** [comment]
Yes, as stated above, the patch is well covered by the testsuite.

Regarding the "bug": I wanted to retain the exact same behavior as we had before. I think you know enough about what we're trying to achieve here for the following to make sense: Let's say we encounter an edge that has not been oriented yet as part of another set of parallel edges. Then we can freely choose how to orient it. The old code had a quirk in the way it did that, in that it chose an entirely unintuitive -- but completely valid -- direction for such edges. If I were to implement things from scratch, I would take a more intuitive, but equally arbitrary direction for these edges. The point is that it really doesn't matter what we do -- so I chose to keep things backward compatible. The comment you refer to simply explains *why* I choose this particular direction.

**Comment by bangerth** [comment]
I've made the changes you requested, and the testsuite is running. Will push these changes once done.

**Comment by tjhei** [comment]
> Regarding the "bug"

Oh, okay. Thanks for explaining. The word "bug" made me think it was incorrect before and you fixed it now.

> I think you know enough about what we're trying to achieve here

Yes, when browsing the code it made sense to me, but I am not familiar enough with it to confirm that it is correct. :-) Anyways, I am happy to have it merged...

**Comment by bangerth** [comment]
Oh, too early, @masterleinad -- I still had @tjhei 's requested changes. Will upload as a separate pull request then.

**Comment by bangerth** [comment]
The rest is in #3171.

**Comment by masterleinad** [comment]
Oh, sorry @bangerth! I've just seen `Ready to merge` and "All checks have passed" and made the conclusion that this can be merged.

**Comment by bangerth** [comment]
Part of #3235.

**Pull request merged**